implementations/rust/ockam/ockam_channel/src/secure_channel_encryptor.rs

```diff
@@ -0,0 +1,88 @@
+use crate::{ChannelKeys, SecureChannelError, SecureChannelVault};
+use ockam_core::async_trait;
+use ockam_core::compat::{boxed::Box, vec::Vec};
+use ockam_core::{Any, Encodable, Result, Route, Routed, TransportMessage, Worker};
+use ockam_node::Context;
+use tracing::debug;
+
+pub(crate) struct SecureChannelEncryptor<V: SecureChannelVault> {
+    keys: ChannelKeys,
+    remote_route: Route,
+    vault: V,
+}
+
+impl<V: SecureChannelVault> SecureChannelEncryptor<V> {
+    pub(crate) fn new(keys: ChannelKeys, remote_route: Route, vault: V) -> Self {
+        Self {
+            keys,
+            remote_route,
+            vault,
+        }
+    }
+
+    pub(crate) fn convert_nonce_from_u64(nonce: u64) -> ([u8; 8], [u8; 12]) {
+        let mut n: [u8; 12] = [0; 12];
+        let b: [u8; 8] = nonce.to_be_bytes();
+
+        n[4..].copy_from_slice(&b);
+
+        (b, n)
+    }
+
+    async fn handle_encrypt(
+        &mut self,
+        ctx: &mut <Self as Worker>::Context,
+        msg: Routed<<Self as Worker>::Message>,
+    ) -> Result<()> {
+        debug!("SecureChannel received Encrypt");
+
+        let reply = msg.return_route();
+        let mut onward_route = msg.onward_route();
+        let transport_message = msg.into_transport_message();
+        let payload = transport_message.payload;
+
+        let _ = onward_route.step();
+
+        let msg = TransportMessage::v1(onward_route, reply, payload.to_vec());
+        let payload = msg.encode()?;
+
+        let payload = {
+            let nonce = self.keys.nonce;
+
+            if nonce == u64::MAX {
+                return Err(SecureChannelError::InvalidNonce.into());
+            }
```

Comment:
Is `0` a valid nonce?

Reply:
Any nonce value is valid, but it's important that unique (key, nonce) pair is used for encryption only once

Reply:
Yes, that makes sense. I'm just confused why the explicit check for u64::MAX